specs/_features/whisk/beacon-chain.md

```diff
@@ -0,0 +1,468 @@
+    whisk_shuffle_proof_M_commitment: BLSG1Point  # [New in Whisk]
+    whisk_registration_proof: ByteList[WHISK_MAX_OPENING_PROOF_SIZE]  # [New in Whisk]
+    whisk_tracker: WhiskTracker  # [New in Whisk]
+    whisk_k_commitment: BLSG1Point  # [New in Whisk]
+```
+
+```python
+def get_shuffle_indices(randao_reveal: BLSSignature) -> Sequence[uint64]:
+    """
+    Given a `randao_reveal` return the list of indices that got shuffled from the entire candidate set
+    """
+    indices = []
+    for i in WHISK_VALIDATORS_PER_SHUFFLE:
+        # XXX ensure we are not suffering from modulo bias
+        shuffle_index = uint256(hash(randao_reveal + uint_to_bytes(i))) % WHISK_CANDIDATE_TRACKERS_COUNT
+        indices.append(shuffle_index)
+
+    return indices
+```
+
+```python
+def whisk_process_shuffled_trackers(state: BeaconState, body: BeaconBlockBody) -> None:
+    # Check the shuffle proof
+    shuffle_indices = get_shuffle_indices(body.randao_reveal)
+    pre_shuffle_trackers = [state.whisk_candidate_trackers[i] for i in shuffle_indices]
+    post_shuffle_trackers = body.whisk_post_shuffle_trackers
+
+    shuffle_epoch = get_current_epoch(state) % WHISK_EPOCHS_PER_SHUFFLING_PHASE
+    if shuffle_epoch + WHISK_PROPOSER_SELECTION_GAP + 1 >= WHISK_EPOCHS_PER_SHUFFLING_PHASE:
+        # Require unchanged trackers during cooldown
+        assert pre_shuffle_trackers == post_shuffle_trackers
+    else:
+        # Require shuffled trackers during shuffle
+        assert IsValidWhiskShuffleProof(
+            pre_shuffle_trackers,
+            post_shuffle_trackers,
+            body.whisk_shuffle_proof_M_commitment,
+            body.whisk_shuffle_proof,
+        )
+
+    # Shuffle candidate trackers
+    for i, shuffle_index in enumerate(shuffle_indices):
+        state.whisk_candidate_trackers[shuffle_index] = post_shuffle_trackers[i]
+```
+
+```python
+def is_k_commitment_unique(state: BeaconState, k_commitment: BLSG1Point) -> bool:
+    return all([validator.whisk_k_commitment != k_commitment for validator in state.validators])
+```
+
+```python
+def process_whisk(state: BeaconState, body: BeaconBlockBody) -> None:
+    whisk_process_shuffled_trackers(state, body)
+
+    # Overwrite all validator Whisk fields (first Whisk proposal) or just the permutation commitment (next proposals)
+    proposer = state.validators[get_beacon_proposer_index(state)]
+    if proposer.whisk_tracker.r_G == BLS_G1_GENERATOR:  # first Whisk proposal
+        assert body.whisk_tracker.r_G != BLS_G1_GENERATOR
+        assert is_k_commitment_unique(state, body.whisk_k_commitment)
+        assert IsValidWhiskOpeningProof(
+            body.whisk_tracker,
+            body.whisk_k_commitment,
+            body.whisk_registration_proof,
+        )
+        proposer.whisk_tracker = body.whisk_tracker
+        proposer.whisk_k_commitment = body.whisk_k_commitment
+    else:  # next Whisk proposals
+        assert body.whisk_registration_proof == WhiskTrackerProof()
```

Comment:
`WhiskTrackerProof` is undefined. It seems we should define a custom type `WhiskTrackerProof: ByteList[WHISK_MAX_OPENING_PROOF_SIZE]`?

Reply:
Yes